### PR TITLE
Move host (formerly Endpoint) to client

### PR DESCRIPTION
**Pull request by jnunemaker**
i have no idea what i am doing

But I think this would make the reading and writing of the host (formerly Endpoint) safe.

### CODE REVIEW

**Comment by joshk** [comment]
WHAT ARE YOU DOING PLAYING WITH GO?? :)

**Comment by joshk** [comment]
Do you think it is overkill to use a Lock? And are other libs using this sort of pattern?

**Comment by jnunemaker** [comment]
LOLOLOL. OH HAI JOSH.

I copied the idea from the ServerList in gomemcached:
https://github.com/bradfitz/gomemcache/blob/master/memcache/selector.go#L37-L84
https://github.com/bradfitz/gomemcache/blob/master/memcache/memcache.go#L116-L117

It seems to me that one would only set the host and then never change it, which makes me think the lock is overkill. BUTTTTTT, if someone did change it, chaos would ensue. I can remove it. I do not feel passionate either way.

**Comment by alindeman** [comment]
> It seems to me that one would only set the host and then never change it, which makes me think the lock is overkill. BUTTTTTT, if someone did change it, chaos would ensue. I can remove it. I do not feel passionate either way.

I support removing it and assuming it won't change as it's being used ... and compel folks do their own locking or serialization if they do.

To me, it seems like the 80% case (maybe even the 99% case?) is creating a client and never changing the host.

I might make it `Host` (capital) to point out that it's accessible.

**Comment by jnunemaker** [comment]
Lock removed. Changed host to Host.

**Comment by jnunemaker** [comment]
In addition, I would move to do the same thing to secure or even just change it to SetScheme and default to http or something, but I can do that in a separate pull.

**Comment by jnunemaker** [comment]
I have also updated the formerly outdated title of the pull request since the lock is gone.

**Comment by timonv** [comment]
Alindeman:

> I support removing it and assuming it won't change as it's being used ... and compel folks do their own locking or serialization if they do.

+5. F me, more people should do this.

jnunemaker:

> I have also updated the formerly outdated title of the pull request since the lock is gone.

<3

> In addition, I would move to do the same thing to secure or even just change it to SetScheme and default to http or something, but I can do that in a separate pull.

This pull would be a _great_ place for that. The party is here!

File: client.go [hunk 1]
```diff
@@ -62,7 +61,17 @@ func (c *Channel) String() string {
...
+	return &Client{
+		appid: appid,
+		key: key,
+		secret: secret,
+		secure: secure,
+		Host: "api.pusherapp.com",
+	}
+}
+
+func (c *Client) SetHost(host string) {
```
Comment (joshk) [inline client.go]: this isn't needed anymore is it as host is now accessible?

Comment (jnunemaker) [inline client.go]: Fixed in b32abae68782d90305be834dfd1c630b73fc5831

**Comment by joshk** [comment]
You :metal: my good sir!

**Pull request merged**